core/src/main/scala/org/http4s/HttpException.scala

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+
+import java.io.IOException
+
+/** Indicates a general failure communicating over HTTP */
+class HttpException(final val message: String = "", final val cause: Option[Throwable] = None)
+    extends IOException(message) {
+  cause.foreach(initCause)
+}
+
+/** Indicates a failure to obtain a response to a request */
+class ResponseException[F[_]](
+    final val request: Request[F],
```

Comment:
Consuming this request body again is probably dangerous, which might make this an outstanding use case for our new `RequestPrelude`.

Reply:
Or perhaps we don't need a custom `ResponseException` at all.  Wrapping in a standard `IOException` would get a trace, let us describe the error, and anyone who wants more from the request can always create a client middleware.

Reply:
Went this direction with it.